Streamline the UI a bit

hoylemd opened the pull request:
refs #41 

While working on #40, (and, well, learning how 2 lua), I made some changes that I think improve the user and development experience.

I haven't run the faketorio tests yet, mainly because I don't know how, and didn't really have internet access while working on these, so I suppose I should probably do that too.

Anyways, run down of the changes:

Main changes:
- Localize the main todo button
- Add explicit styles for UI elements, and use them
- Change the 'show button when minimized' setting to just always show the button
- Main Todo List button now toggles the UI, just like the hotkey
- Unify add and edit UIs
- Move add/edit UI init into UI module

QOL:
- ~todo.log also logs to main log~ Removed because lint doesn't like it and I don't think I can change the travis config to allow it. Also, not really super helpful.

fixes:
- Add chinese localization entries for auto-assign (in english, but at least they're there?)
- squelch the 'unknown element' warning when a non-todo gui element is clicked.

See in-line github comments for some more detailed explanations. (still coming)

Screenshots!

Before:
<img width="954" alt="screenshot 2018-08-06 16 04 39" src="https://user-images.githubusercontent.com/569403/43738652-5bd7ca2c-9993-11e8-9a9c-b6cb0ebe7b28.png">

After:
<img width="1140" alt="screenshot 2018-08-06 15 51 03" src="https://user-images.githubusercontent.com/569403/43738700-876ac07c-9993-11e8-978c-f145840f4413.png">

JonasJurczok commented:
Hi,
thanks for the PR.

Regarding the faketorio tests.. ideally it should be as easy as running `luarocks install faketorio` and then a `faketorio test` in the mod folder. Unfortunately there seems to be a problem with installing faketorio on windows (I was doing all developement on Linux until recently and now have to get my win environment to work).

I like the new styles. This also fixed the line break problem. nice :)

I'm not sure about keeping the button there all the time. The initial idea was to save that screenspace to be able to show other buttons etc. But tbh I play with the button invisible anyway.

I'll definitely take a look at your changes.

hoylemd commented:
I'm working on macos, actually.  I was able to get faketorio working, but it took some tinkering, and a lot of the functionality didn't really work (e.g. it couldn't create or find the new save when running `faketorio test`, but I just started a new game and ran `/faketorio` and it worked.  There's only supposed to be the one feature, right?

l updated the label for the button setting so that it just says 'show the button in the mod ui'.  That way, you can still play with the invisible button, but if you're like me, you can have it persist.  I see a lot of other mods do this too (helmod and outpost planner are 2 examples).  Also updated the name of the setting to reflect it's functionality.

I updated the french and german localizations too, but my french and german aren't *great*, so if they seem weird to a native speaker please suggest an improvement.  Thankfully, the longer descriptions seemed to already describe the functionality I was going for, so I only had to modify the short ones.

I'm also not 100% sure I updated the faketorio test correctly, and given that CI is failing, I'm thinking I didn't.  So I'll check that out next.

hoylemd commented on src/control.lua:
```diff
@@ -22,16 +22,15 @@ end)
+
+require("faketorio.core")
+
+require("faketorio.features.UI_feature")
```
I don't know why this is missing on master now, I think it was in master when I started this branch?  Seems to be important for running faketorio tests ¯\_(ツ)_/¯

JonasJurczok replied:
This should not be in master. Because if it is there then you can run faketorio tests, but it also requires everyone to have faketorio installed. One of the features of faketorio is to generate these lines for you when you run the `test` command :)

JonasJurczok replied:
As already said in the main thread: This should not be here otherwise it won't work without faketorio anymore. Please remove it :D

hoylemd replied:
Ah! that explains why the lines suddenly appeared. Thanks, will do.

hoylemd replied:
Actually, it does work without faketorio (at least on windows).  I'm impatient, so I packaged it all up and put the .zip on my windows gaming machine (where I don't have faketorio installed) and it worked great for a few hours last night.  Just FYI - still planning to remove it.

hoylemd commented on src/todo/UI.lua:
```diff
@@ -2,11 +2,14 @@
-    if not todo.get_maximize_button(player) and not todo.get_main_frame(player) and todo.show_minimized(player) then
+    if (not todo.get_maximize_button(player)
+      and not todo.get_main_frame(player)
+      and todo.show_button(player) ) then
```
I thought this was easier to read, but maybe just me.

JonasJurczok replied:
Works for me :)

hoylemd commented on src/todo/UI.lua:
```diff
@@ -121,40 +135,47 @@ function todo.create_add_edit_frame(player)
-        name = "todo_new_task_textbox"
+        style = "todo_textbox_default",
+        name = "todo_new_task_textbox",
+        text = task and task.task or ""
```
I don't know if this is common in lua, but this is sort of a 'poor man's ternary'.

If task is truthy, then the result of the expression will be task.task. Otherwise, emptystring. Analogous to

text = task ? task.task : ""; in javascript.

(if this is already a well-known pattern, sorry. I'm very new to lua :p)

I used this in a few places, not just here.

JonasJurczok replied:
I'm also new to this :) If it works fine. I think the usage of and and or is extremely unintuitive, though.

hoylemd replied:
Yeah, that's the downside of this pattern.  I'm thinking about just putting it in an if block anyways.  This is really nice and concise, but readability is more important.  I'll see what looks better.

hoylemd commented on src/todo/UI.lua:
```diff
@@ -62,51 +68,59 @@ function todo.create_task_table(frame, player)
-function todo.create_add_edit_frame(player)
+function todo.create_add_edit_frame(player, task)
     local gui = player.gui.center
+    task = task or nil
```
Not sure if this is really necessary, but better safe than sorry, I guess?

JonasJurczok replied:
I did an experiment once and it should not be necessary, but I cannot recall the details anymore. So I aggree.. better safe than sorry :D

hoylemd commented on src/todo/UI.lua:
```diff
@@ -121,40 +135,47 @@ function todo.create_add_edit_frame(player)
+    if task then
+        assign_index = task.assignee and lookup[task.assignee] or 1
+    else
+        if( todo.is_auto_assign(player) and c == 1 ) then
```
Hmm, would it be better to merge this with the `else` to form an `elseif`?

JonasJurczok replied:
yes

hoylemd commented on src/todo/todo.lua:
```diff
@@ -96,32 +99,6 @@ function todo.create_task(text, assignee)
-        caption = {"todo.update"}
-    })
-end
-
```
This was merged with the `create_add_edit` function in UI

JonasJurczok replied:
I'm always a fan of deleting code :)

hoylemd commented:
Alrighty! Finally ready for review, whenever you have the time. I've also got a 'move to top/bottom' branch working, but it needs more polish before a PR.  Maybe later this week :)

JonasJurczok commented:
Awesome :)

Can you please report the problems with faketorio [here](https://github.com/JonasJurczok/faketorio/issues)? Then I'm aware of them and can try to reproduce/fix them directly.

JonasJurczok commented on locale/de/locale.cfg:
```diff
@@ -23,13 +24,13 @@ hide_done=Verberge abgeschlossene Aufgaben
 
 [mod-setting-name]
-todolist-show-minimized=Zeige "Todo List" Schalter wenn minimiert.
+todolist-show-button=Zeige "Todo List" Schalter ins der mod ui.
```
I suggest
`Zeige "Todo List" Schalter an/verstecke ihn`

hoylemd replied:
If I'm reading that correctly (with my rather poor german augmented by google translate), it seems to translate to 'Show/Hide "Todo List" button'.  Maybe we just make them all say their equivalent of 'Show "Todo List" button', without mentioning hiding or the location.

I'm thinking that, because it's a checkbox/boolean value, the label should be descriptive of the 'on' state, with the state of the checkbox providing the polarity(?) of the setting (as in checked = 'this statement is true', unchecked = 'negate this statement')

JonasJurczok commented on spec/UI_feature.lua:
```diff
@@ -20,8 +20,8 @@ feature("Testing the UI", function()
     end)
-end)
\ No newline at end of file
+end)
```
looks good.

JonasJurczok commented on src/control.lua:
```diff
@@ -22,16 +22,15 @@ end)
-        todo.maximize(player)
-        todo.refresh_task_table(player)
-    end
+    todo.toggle_ui(player)
```
Nice

JonasJurczok commented on src/todo/style.lua:
```diff
@@ -0,0 +1,45 @@
+local default_gui = data.raw["gui-style"].default
```
Nice. I never got around to learn how styles work in Factorio. Thanks for implementing this :)

hoylemd replied:
Too many years of front-end web development, finally paying off!

JonasJurczok commented on src/todo/todo.lua:
```diff
@@ -275,23 +251,23 @@ function todo.on_gui_click(event)
-    else
-        todo.log("Unknown element name:" .. element.name)
+    elseif (string.find(element.name, "todo_")) then
+        todo.log("Unknown todo element name:" .. element.name)
```
+1

hoylemd commented:
The final commit is just my suggestion for a more concise wording of the german option.  If @JonasJurczok Would prefer the show/hide phrasing, then I'll update all 3 roman-script localizations to the best of my ability (my french is better than my german. Yay Canadian education system!).  Someone who knows chinese will have to do that one though.

As for the faketorio issues, I'll post an issue, but It probably won't be super detailed.  I don't remember a lot of the specifics.

Tarrke commented:
Please no Canadian French. Please please please. Promise I won't translate to English.

**joke intended**

hoylemd commented:
@Tarrke  Oh, don't worry, the Ontario(my province) curriculum actually does teach *real* french, not the Quebecois pidgin :p

But I'm far from fluent - please do correct my bad french :p

(also thank you for specifying intent - I probably would have misinterpreted otherwise. )

hoylemd commented:
Huh, I don't know what's up with that CI error.  Maybe it just needs to be run again?  Couldn't find a button for that :/

hoylemd commented:
hoorj it worked!

hoylemd commented:
@JonasJurczok Hey, I don't mean to be a bother, but do you think you could re-review the PR today or tomorrow?

JonasJurczok commented:
Yes, sorry.. I got distracted with the faketorio issues on my windows
machine (development on windows is..........), Also I was barely at home
the last days.

I'll try to get it done tomorrow, otherwise Saturday.

Thanks for your work :)

Michael D. Hoyle <notifications@github.com> schrieb am Do., 9. Aug. 2018,
18:37:

> @JonasJurczok <https://github.com/JonasJurczok> Hey, I don't mean to be a
> bother, but do you think you could re-review the PR today or tomorrow?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JonasJurczok/factorio-todo-list/pull/44#issuecomment-411820443>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFsdP3siqw1DpFu-tZUms3tMLaUyy-pks5uPGVWgaJpZM4VxBB8>
> .
>

hoylemd commented:
:+1: Totally cool.  I feel you on the windows development. I have to use Azure a lot for work and it's..............

hoylemd commented:
hoylemd is my handle!  But if you prefer real names, Mike Hoyle is good too.

Thanks! Now to rebase and PR my 'move to top/bottom' branch :p

hoylemd commented:
Huh, the Travis build for the merge failed with that `busted could not be found` error again. I saw that while working on this branch. I wonder what's causing that?

JonasJurczok commented:
I assume network error on Travis side. I planned to add all the luarocks
directory to the cache anyway. This would prevent it from downloading
everything again and again. Overall it should be fine though as the pr also
builds the merge result.

hoylemd commented:
Yeah, I knew about the merge bit.  Just a little annoying.  Adding packages to the cache is a good idea though!
